docs/contributing/coding-style.md

```diff
@@ -923,34 +934,83 @@ Still, in some cases, you may need to change the generated code to meet your req
+   The generated test code will default use [goleak](https://github.com/uber-go/goleak) library to test if there is any goroutine leak.
+   Sometimes you may want to skip the detection; for example, Vald uses [fastime](https://github.com/kpango/fastime) library, but the internal goroutine is not closed due to the needs of the library.
    To skip the goleak detection, we need to create the following variable to store the ignore function.
 
    ```go
    var (
-       // Goroutine leak is detected by `fastime`, but it should be ignored in the test because it is an external package.
+       // goroutine leak is detected by `fastime`, but it should be ignored in the test because it is an external package.
        goleakIgnoreOptions = []goleak.Option{
            goleak.IgnoreTopFunction("github.com/kpango/fastime.(*fastime).StartTimerD.func1"),
        }
    )
    ```
 
-   And modify the generated test code.
+   And use it in `VerifyNone()` in test case.
 
    ```go
-   // before
-   for _, test := range tests {
-       t.Run(test.name, func(tt *testing.T) {
-           defer goleak.VerifyNone(tt)
-
-   // after
-   for _, test := range tests {
+   for _, tc := range tests {
+       test := tc
        t.Run(test.name, func(tt *testing.T) {
            // modify the following line
            defer goleak.VerifyNone(tt, goleakIgnoreOptions...)
    ```
 
+   In Vald, we implemented [internal goleak package](https://github.com/vdaas/vald/blob/main/internal/test/goleak/goleak.go) and wrap the goleak validation logic to ignoring the common goleak functions in Vald use case.
+   In test implementation, we can simplily import the internal goleak package and ignore all the necessary goleak function by default.
+
+   ```go
+   // import internal goleak package
+   import (
+       "github.com/vdaas/vald/internal/test/goleak"
+   )
+
+   ...
+   for _, tc := range tests {
+       test := tc
+       t.Run(test.name, func(tt *testing.T) {
+           // use internal goleak to ignore necessary function by default
+           defer goleak.VerifyNone(tt, goleak.IgnoreCurrent())
+   ```
+
+1. goleak usage
+
+   There are two methods for valid goroutine leak:
+
+   1. Use `goleak.VerifyNone()` to validate it on each test cases.
+   2. Use `goleak.VerifyTestMain()` to validate it on each package.
+
+   By default in Vald, the goroutine leak validation is executed in each test case. e.g.
+
+   ```go
+   for _, tc := range tests {
+        test := tc
+        t.Run(test.name, func(tt *testing.T) {
+            defer goleak.VerifyNone(tt, goleak.IgnoreCurrent())
+   ```
+
+   In some cases, it may not work as some cleanup process or asynchronous process remaining in the background, and validating it on the test case using `goleak.VerifyNone()` may cause a false alarm.
+
+   To resolve it, we can consider using `goleak.VerifyTestMain()` to validate goleak when all test cases are passed in each packages, to avoid the goleak false alarm.
```

Comment:
```suggestion
   To resolve it, we can consider using `goleak.VerifyTestMain()` to validate goleak when all test cases are passed in each package, to avoid the goleak false alarm.
```